cis194/week6/stevemao/Fibonacci.hs

```diff
@@ -0,0 +1,55 @@
+{-# OPTIONS_GHC -Wall #-}
+module Fibonacci where
+
+fib :: Integer -> Integer
+fib 0 = 0
+fib 1 = 1
+fib n = (fib $ n - 1) + (fib $ n - 2)
+
+fibs1 :: [Integer]
+fibs1 = fmap fib [0..]
+
+fib' :: Integer -> Integer
+fib' 0 = 0
+fib' 1 = 1
+fib' n = snd $ foldr (\_ acc -> (snd acc, fst acc + snd acc)) (0, 1) [2..n]
+
+fibs2 :: [Integer]
+fibs2 = fmap fib' [0..]
+
+data Stream a = Cons a (Stream a)
+
+streamToList :: Stream a -> [a]
+streamToList (Cons a s) = (a : streamToList s)
+
+instance Show a => Show (Stream a) where
+  show s = show . take 20 $ streamToList s
+
+streamFromList :: [a] -> Stream a
+streamFromList (x: xs) = Cons x $ streamFromList xs
+streamFromList _ = error "List must not be empty"
+
+streamRepeat :: a -> Stream a
+streamRepeat a = streamFromList $ repeat a
+
+streamMap :: (a -> b) -> Stream a -> Stream b
+streamMap f s = streamFromList . fmap f . streamToList $ s
+
+streamFromSeed :: (a -> a) -> a -> Stream a
+streamFromSeed f = streamFromList . iterate f
+
+nats :: Stream Integer
+nats = streamFromList [0..]
```

Comment:
Because `steamFromList` needs to be changed to have a type that does not fit here, we can utilise `streamFromSeed` above.